### DISCUSSION TITLE
What is the difference between `getInitialProps` and `getServerSideProps`?

### DISCUSSION THREAD

**Discussion by chaitanyaj2222** [vercel/next.js#47357] [Help] [answered]
### Summary

I learned about both functions from the documentation. I couldn't notice any difference between them. Could someone tell me how both of them differ?

**Comment by icyJoseph** [comment, marked as answer]
`GetServerSideProps` (GSSP), runs exclusively server side, while `getInitialProps`, runs initially server side and then on client transitions, runs client side, unless paired with a page which uses GSSP.

The thing with `GSSP`, is that when you do client side transitions, using next/router or next/link, the framework calls for a `_next/data` endpoint, which hits your GSSP function, again, server side, and it collects the results as JSON, which is transferred to the client, and used as props for your page component.

So `GSSP` presents us with a very powerful abstraction, the ability to generate props for a component server side. It has a pitfall though, that it might be slow, because it must wait for the GSSP function to complete.

When a user arrives to your application, at a page using `GSSP`, the framework generates all of the HTML for that page on demand.

`GSSP` is an improvement on GetInitiaProps, by specialisation. Instead of being a hybrid, it is exclusive to the server side, and pages which use it, are generated on-demand, or during SPA transitions (with next/router, next/link), obtain their props from a network call.

**Reply by chaitanyaj2222** [reply]
I really loved the way you explained the differences between both the functions. Thank you for your valuable time and knowledge. Could you please clarify a couple more doubts I have with your explanation?

1. Reloading a page that uses `getInitialProps` runs this function in the browser itself and renders the page with new props on the client-side. Is this correct?

2. But, reloading a page that uses `GSSP` runs this function on the _server side_ and returns the props to the _client_ and then the page is rendered on the _client side_ using the new props. Is it correct or do we get the _pre-rendered page from server side with new props_?

Ps: By **reloading**, I am also talking about navigating to the page using `next/link` or `next/router`

**Reply by icyJoseph** [reply]
The thing is that historically, getInitialProps was the way to get data on Next.js up to version 9.x.x (don't remember exactly which), which introduced GetStaticProps, GSSP, and GetStaticPaths.

So these three were a refinement on the GIP idea.

> Reloading a page that uses getInitialProps runs this function in the browser itself and renders the page with new props on the client-side. Is this correct?

No, it won't run on the browser after page reload.

Reloading the page is, for the most part, like opening a new tab and requesting that page again. Everything that lives in the HTML/CSS/JavaScript context is destroyed and redone.

So you'd run GIP server side, unless the page was optimised, which would happen if the page uses GetStaticProps, then it won't run at all.

> For the initial page load, getInitialProps will run on the server only. getInitialProps will then run on the client when navigating to a different route via the [next/link](https://nextjs.org/docs/api-reference/next/link) component or by using [next/router](https://nextjs.org/docs/api-reference/next/router). However, if getInitialProps is used in a custom _app.js, and the page being navigated to implements getServerSideProps, then getInitialProps will run on the server.

GIP requires a bit more cognitive overhead... which is why they introduced specialised versions. Lay GetStaticProps on top and you are in for a ride... 

> But, reloading a page that uses GSSP runs this function on the server side and returns the props to the client and then the page is rendered on the client side using the new props. Is it correct or do we get the pre-rendered page from server side with new props?

Nope, this would, again, run server side totally. Refreshing a page unloads it whole.

**Reply by chaitanyaj2222** [reply]
Got it @icyJoseph.

Reloading is just like visiting the website for the very first time. So, GSSP will run on the server side. Cool!

>The thing with GSSP, is that when you do client side transitions, using next/router or next/link, the framework calls for a _next/data endpoint, which hits your GSSP function, again, server side, and it collects the results as JSON, which is transferred to the client, and used as props for your page component.

I tested this using browser dev tools I was able to see that client side page transition to a page that uses GSSP fetches only the props. I did not know that Next.js does this. I used to think that the framework will get the _whole pre-rendered page_ every time while navigating to a page that uses GSSP. It's now clear to me that only for the _first time_ do we get the pre-rendered page but every other client-side transition to the same page will _only fetch the props_ needed to be passed to the component.  Please correct me if I got it wrong.
